src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,258 @@
+            _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        public abstract TextDocumentIdentifier? GetTextDocumentIdentifier(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Gets the progress object to stream results to.
+        /// </summary>
+        protected abstract IProgress<TReport[]>? GetProgress(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Retrieve the previous results we reported.  Used so we can avoid resending data for unchanged files. Also
+        /// used so we can report which documents were removed and can have all their diagnostics cleared.
+        /// </summary>
+        protected abstract DiagnosticParams[]? GetPreviousResults(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Returns all the documents that should be processed in the desired order to process them in.
+        /// </summary>
+        protected abstract ImmutableArray<Document> GetOrderedDocuments(RequestContext context);
+        protected abstract TReport CreateReport(TextDocumentIdentifier? identifier, VSDiagnostic[]? diagnostics, string? resultId);
+
+        private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateArgs)
+        {
+            if (updateArgs.DocumentId == null)
+                return;
+
+            // Whenever we hear about changes to a document, drop the data we've stored for it.  We'll recompute it as
+            // necessary on the next request.
+            _documentIdToLastResultId.Remove((updateArgs.Workspace, updateArgs.DocumentId));
+        }
+
+        public async Task<TReport[]?> HandleRequestAsync(
+            TDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
+        {
+            // The progress object we will stream reports to.
+            var progress = GetProgress(diagnosticsParams);
+
+            // The buffer we will add results to if our client doesn't support streaming results.
+            using var _ = ArrayBuilder<TReport>.GetInstance(out var reports);
+
+            // Get the set of results the request said were previously reported.  We can use this to determine both
+            // what to skip, and what files we have to tell the client have been removed.
+            var previousResults = GetPreviousResults(diagnosticsParams) ?? Array.Empty<DiagnosticParams>();
+
+            var documentToPreviousResult = new Dictionary<Document, DiagnosticParams>();
+            foreach (var previousResult in previousResults)
+                AddPreviousResult(documentToPreviousResult, previousResult);
+
+            // First, let the client know if any workspace documents have gone away.  That way it can remove those for
+            // the user from squiggles or error-list.
+            HandleRemovedDocuments(previousResults, progress, reports);
```

Comment:
Not that its super complicated this way, but `IProgress` is a pretty simple interface, it would be straightfoward to write a class that implemented it and just collected results, so that `progress` and `reports` don't both need to be passed everywhere

Reply:
yup!  i'm in agreement.  i don't like thsi approach, and we should create a reusable pattern for this.